### PR TITLE
Add more testing for IMAP headers in sent folder

### PR DESCRIPTION
**Pull request by quynhn**

### CODE REVIEW

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 1]
```diff
@@ -2327,4 +2327,198 @@ public void attachmentAndEmptyBodyShouldBeRetrievedWhenChainingSetMessagesAndGet
...
+                    hasEntry(equalTo("Message-ID"), not(isEmptyOrNullString())),
+                    hasEntry(equalTo("MIME-Version"), not(isEmptyOrNullString())),
+                    hasEntry(equalTo("Subject"), not(isEmptyOrNullString())),
+                    hasEntry(equalTo("Date"), not(isEmptyOrNullString())),
+                    hasEntry(equalTo("Content-Type"), not(isEmptyOrNullString()))
+                )))
+            ;
+            return true;
+        } catch(AssertionError e) {
+            e.printStackTrace();
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Throwables.propagate

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 2]
```diff
@@ -2327,4 +2327,198 @@ public void attachmentAndEmptyBodyShouldBeRetrievedWhenChainingSetMessagesAndGet
...
+
+        given()
+            .header("Authorization", accessToken.serialize())
+            .body(requestBody)
+            .when()
+            .post("/jmap")
+            .then()
+            .log().ifValidationFails()
+            .statusCode(200)
+            .body(NAME, equalTo("messagesSet"))
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Those assertions are not needed in this test.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 3]
```diff
@@ -2327,4 +2327,198 @@ public void attachmentAndEmptyBodyShouldBeRetrievedWhenChainingSetMessagesAndGet
...
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: same comment on assertions

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 4]
```diff
@@ -2327,4 +2327,198 @@ public void attachmentAndEmptyBodyShouldBeRetrievedWhenChainingSetMessagesAndGet
...
+                hasEntry(equalTo("isAnswered"), equalTo(false))
+            )))
+        ;
+        String sentMailboxId = getMailboxId(accessToken, Role.SENT);
+
+        calmlyAwait.atMost(60, TimeUnit.SECONDS).until( () -> messageOnSentBoxHasHeaders(sentMailboxId));
+
+    }
+
+    private boolean messageOnSentBoxHasHeaders(String sentMailboxId) {
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: You may reuse previous method (messageOnInboxHasHeaders), just add a parameter for the mesage Id.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 5]
```diff
@@ -2327,4 +2327,198 @@ public void attachmentAndEmptyBodyShouldBeRetrievedWhenChainingSetMessagesAndGet
...
+                    hasEntry(equalTo("Message-ID"), not(isEmptyOrNullString())),
+                    hasEntry(equalTo("MIME-Version"), not(isEmptyOrNullString())),
+                    hasEntry(equalTo("Subject"), not(isEmptyOrNullString())),
+                    hasEntry(equalTo("Date"), not(isEmptyOrNullString())),
+                    hasEntry(equalTo("Content-Type"), not(isEmptyOrNullString()))
+                )))
+            ;
+            return true;
+        } catch(AssertionError e) {
+            e.printStackTrace();
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Throwables.propagate

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 6]
```diff
@@ -2327,4 +2327,198 @@ public void attachmentAndEmptyBodyShouldBeRetrievedWhenChainingSetMessagesAndGet
...
+            "      }}" +
+            "    }," +
+            "    \"#0\"" +
+            "  ]" +
+            "]";
+
+        given()
+            .header("Authorization", accessToken.serialize())
+            .body(requestBody)
+            .when()
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Ident on when and then to improve readability
(on intelliJ you need to manually delete 4 spaces before the dot)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 1 above)
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Nope, but please remove the noisy stacktrace ... If we have troubles with this test, we will add it or add logs on JMAP request.
Could you remove the printStackTrace ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 5 above)
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Idem

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 2 above)
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Done

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 1 above)
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Done

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 3 above)
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Removed

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 4 above)
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Done

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 5 above)
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Done

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 7]
```diff
@@ -2361,67 +2361,12 @@ public void setMessageShouldVerifyHeaderOfMessageOnInbox() throws Exception {
...
-                    hasEntry(equalTo("Date"), not(isEmptyOrNullString())),
-                    hasEntry(equalTo("Content-Type"), not(isEmptyOrNullString()))
-                )))
-            ;
-            return true;
-        } catch(AssertionError e) {
-            e.printStackTrace();
-            return false;
-        }
+        calmlyAwait.atMost(60, TimeUnit.SECONDS).until( () -> messageOnMailboxHasHeaders(inboxMailboxId, "username1@domain.tld|INBOX|1"));
```
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: messageInMailbox ?

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java (hunk 6 above)
Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Not done

Comment (chibenwa) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: You are supposed to have : 

```
given()
    .header()
    .body()
.when()
    .anyThing()
.then()
    .check()
```

but you have : 


```
given()
    .header()
    .body()
    .when()
    .anyThing()
    .then()
    .check()
```

Like other calls like this in this class

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 8]
```diff
@@ -2327,4 +2327,124 @@ public void attachmentAndEmptyBodyShouldBeRetrievedWhenChainingSetMessagesAndGet
...
+            .statusCode(200)
+            .body(ARGUMENTS + ".created", aMapWithSize(1))
+        ;
+        String sentMailboxId = getMailboxId(accessToken, Role.SENT);
+
+        calmlyAwait.atMost(60, TimeUnit.SECONDS).until( () -> messageInMailboxHasHeaders(sentMailboxId, "username@domain.tld|sent|1"));
+
+    }
+
+    private boolean messageInMailboxHasHeaders(String mailboxId, String messageId) {
```
Comment (mbaechler) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: this method should take the list of expected headers.

Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Done

**Comment by quynhn** [comment]
>This issue should have a matching Jira and the commit message should be modified to include it. 

Done

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java [hunk 9]
```diff
@@ -2408,11 +2410,23 @@ public void setMessageShouldVerifyHeaderOfMessageInSent() throws Exception {
...
+    private ImmutableMap buildExpectedHeaders() {
+        ImmutableMap.Builder expectedHeaders = ImmutableMap.builder();
+        expectedHeaders.put("sender", "Me <username@domain.tld>");
+        expectedHeaders.put("encoding", "quoted-printable");
+        expectedHeaders.put("from", "Me <username@domain.tld>");
+        expectedHeaders.put("to", "BOB <username1@domain.tld>");
+        expectedHeaders.put("version", "1.0");
+        expectedHeaders.put("subject", "Thank you for joining example.com!");
+        expectedHeaders.put("content-type", "text/plain; charset=UTF-8");
+        return expectedHeaders.build();
```
Comment (quynhn) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/SetMessagesMethodTest.java]: Done

**Pull request merged**